lib/pos_remote_node.go

```diff
@@ -0,0 +1,456 @@
+package lib
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/deso-protocol/core/bls"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/sha3"
+	"math"
+	"net"
+	"time"
+)
+
+type RNConnectionType int
+
+const (
+	RNConnectionType_Unknown  RNConnectionType = 0
+	RNConnectionType_Outbound RNConnectionType = 1
+	RNConnectionType_Inboound RNConnectionType = 2
+)
+
+type RNConnectionStatus int
+
+const (
+	RNConnectionStatus_NotConnected RNConnectionStatus = 0
+	RNConnectionStatus_Connected    RNConnectionStatus = 1
+	RNConnectionStatus_Attempted    RNConnectionStatus = 2
+	RNConnectionStatus_Terminated   RNConnectionStatus = 3
+)
+
+//type RemoteNodeId struct {
+//	PeerId    uint64
+//	AttemptId uint64
+//}
+
+type RemoteNode struct {
+	// Should we have ID here?
+	peer             *Peer
+	id               RemoteNodeId
+	connectionType   RNConnectionType
+	connectionStatus RNConnectionStatus
+
+	params *DeSoParams
+
+	bc  *Blockchain
+	srv *Server
+
+	minTxFeeRateNanosPerKB uint64
+	hyperSync              bool
+	posValidator           bool
+
+	handshakeMetadata *HandshakeMetadata
+	keystore          *BLSKeystore
+}
+
+type HandshakeMetadata struct {
+	versionNoncesSent         uint64
+	versionNoncesReceived     uint64
+	userAgent                 string
+	versionNegotiated         bool
+	ServiceFlag               ServiceFlag
+	advertisedProtocolVersion ProtocolVersionType
+	negotiatedProtocolVersion ProtocolVersionType
+	minTxFeeRateNanosPerKB    uint64
+	timeConnected             *time.Time
+	timeOffsetSecs            int64
+	versionTimeExpected       *time.Time
+	verackTimeExpected        *time.Time
+	StartingBlockHeight       uint32
+	validatorPublicKey        bls.PublicKey
+}
+
+func NewRemoteNode() *RemoteNode {
+	return &RemoteNode{
+		id:                NewRemoteNodeNoId(),
+		connectionType:    RNConnectionType_Unknown,
+		connectionStatus:  RNConnectionStatus_NotConnected,
+		handshakeMetadata: nil,
+	}
+}
+
+func (rn *RemoteNode) SetConnectedPeer(peer *Peer) {
+	rn.peer = peer
+	if rn.peer == nil {
+		return
+	}
+
+	// Set connectionType
+	if rn.peer.IsOutbound() {
+		rn.connectionType = RNConnectionType_Outbound
+		rn.SetId(NewRemoteNodeOutboundId(peer.ID, peer.AttemptId()))
+	} else {
+		rn.connectionType = RNConnectionType_Inboound
+		rn.SetId(NewRemoteNodeInboundId(peer.ID))
+	}
+
+	// Set connectionStatus
+	rn.connectionStatus = RNConnectionStatus_Connected
+}
+
+func (rn *RemoteNode) GetPeer() *Peer {
+	return rn.peer
+}
+
+func (rn *RemoteNode) IsInbound() bool {
+	return rn.connectionType == RNConnectionType_Inboound
+}
+
+func (rn *RemoteNode) IsOutbound() bool {
+	return rn.connectionType == RNConnectionType_Outbound
+}
+
+func (rn *RemoteNode) SetId(id RemoteNodeId) {
```

Comment:
I made it private, since it'll really only be accessed inside of RemoteNode. The RemoteNode id can change depending on the lifecycle of the connection. Attempted connection will have (peerId, attemptId) = (0, y) while a formed connection will have (peerId, attemptId) = (x, 0) or (x, y), which will be adjusted over time.